firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java

```diff
@@ -369,6 +371,20 @@ public Direction getKeyOrder() {
     return this.orderBys.get(this.orderBys.size() - 1).getDirection();
   }
 
+  /** Returns the number of segments of a perfect index for this target. */
+  public Integer getSegmentCount() {
+    Set<FieldPath> fields = new HashSet<>();
+    for (Filter filter : filters) {
+      for (FieldFilter subFilter : filter.getFlattenedFilters()) {
+        fields.add(subFilter.getField());
+      }
+    }
+    for (OrderBy orderBy : orderBys) {
+      fields.add(orderBy.getField());
+    }
+    return fields.size();
```

Comment:
I think we need to handle some exclusions:

```
query.where('a', '>', 42).orderBy('a')
```

and

```
query.where('a', '>', 42).where('a', '<', 1337)
```

both have a "perfect" segment count of 1.

Reply:
My assumption was that if an index can be used to serve a target and the number of index segments equals the number of unique FieldPaths in the target's filters&orderBys, then it is the perfect index for it.

I'm a bit concerned whether this is the right assumption; but I cannot think of any case where this is incorrect. Any thoughts?

For the first case `a>42 orderby a`, the "perfect" segment count of 1 sounds good to me (`A ASC`)
For the second case, it has multiple inequalities and I don't know how that should be served from the index. In theory, the `A ASC` can be used two times to compute the results.

Reply:
I think we need to count the number of unique field paths that are used in both `orderBy` and `where`. We do need to special case array-contains and array-contains-any filters, which would require an additional `IndexKind.CONTAINS` segment that is independent from any ordering.